src/main/java/com/google/devtools/build/lib/remote/options/RemoteOptions.java

```diff
@@ -59,13 +59,13 @@
 
   @Option(
       name = "remote_max_connections",
-      defaultValue = "100",
+      defaultValue = "0",
```

Comment:
Isn't 0 for HTTP the same as no limit?

Reply:
Yes. I am actually concerned about changing to 0.

The number of HTTP connections has been limited to 100 by default. By changing to 0, it could grow rapidly for a large build.

On the other hand, the number of gRPC channels was 1 (<`3.7.0`), then changed to based on `--jobs` (< `4.1.0`) and finally changed to dynamic connection pool (>= `4.1.0`) which doesn't has an upper-bound. Note that one gRPC channel could usually handle 100+ concurrent requests. So leaving `--remote_max_connections` to 100 by default means we can handle 10000+ max concurrent requests by default which looks find to me. By building Bazel itself with this change, I didn't see the number of gRPC channels reaching the max.

WDTY if we keep the 100 as default?

Reply:
Since projects ran into issues before this flag was added, which they would possibly encounter again if this became `0`, I say we need to leave it at `100`.

I would adjust the help text to indicate how it affects HTTP versus gRPC, especially since people might not know it now affects the gRPC side. I would be surprised that if this is `100` it would allow "10,000 concurrent requests", as I'm telling it "100 max connections". It would be nice if it spelled out that "connection" == "gRPC channel" as well.

Reply:
Thanks for the input! SGTM.

If we keep the default of `--remote_max_connections` to `100` and make it to the number of max concurrent requests on the fly for gRPC, the number of gRPC channels is limited to `1` by default which could cause problems for some projects. I need to guard this feature with an incompatible flag which could be flipped and removed in `6.x`.

Reply:
Just want to emphasize here: for each gRPC channel, it could create one or more HTTP/2 connections depending on resolved IP. For each connection, it could handle `MAX_CONCURRENT_STREAMS` (which is a per-server settings and is usually set to `100` or `128`) concurrent requests. (See more details in https://github.com/bazelbuild/bazel/issues/11801)

AFAIK, there isn't an API to check the max concurrent requests a channel could handle, so we assume the number is `100`. Then the max number of gRPC channels we can create is calculated by `--remote_max_connections / 100`.

Reply:
I guess I wonder if `--remote_max_connections` _should_ be "max concurrent requests on the fly for gRPC". https://github.com/bazelbuild/bazel/issues/14178's concern is about the number of sockets. I assume each concurrent stream of a gRPC channel opens a socket? If so, then yes, this change (making it `/ 100` for gRPC) makes sense, otherwise I think it should equal the maximum number of channels.

Reply:
> I assume each concurrent stream of a gRPC channel opens a socket? 

Nope.

To make it easier to understand, assuming the server name is resolved to one IP address for the following discussion.

One gRPC channel opens one TCP connection. It can use this one TCP connection to fire off 100+ requests at one time by leveraging HTTP/2 multiplexing. So with the latest commit, if `--remote_max_connections` is set to `100`, Bazel will create only one gRPC channel which opens only one TCP connection but can make 100 concurrent requests.

The semantic of `--remote_max_connections` was to limit max connections for HTTP remote cache. Since we only send one request at one time per HTTP connection, this implies to limiting the max concurrent requests. 

However, since gRPC can more requests at one time per connection, this becomes ambiguous.

1. If we want `--remote_max_connections` limit the max number of TCP connections (or sockets), Bazel could make up to `--remote_max_connections * 100` concurrent requests to the server.
2. If we want `--remote_max_connections` limit the max number of concurrent requests, Bazel could open up to  `--remote_max_connections / 100` TCP connections (or sockets) to the server.

Both seem reasonable to me, but in terms of implementation and backward compatibility, I prefer 1 since we don't need to introduce an incompatible flag (the default value is high enough to consider unlimited for gRPC).

WDTY in your use case?

Reply:
I agree, it should be 1, as "connections" == "sockets" in my mind, not requests.